### ISSUE TITLE
Java memory adjustments in Docker version with XML backend

### ISSUE THREAD

**Issue by JonasZal** [phax/phoss-smp#291]
We have an SMP instance with the XML backend and 3999 (+1200 after a few weeks) service groups. Upon endpoint update, we got "java.lang.OutOfMemoryError: Java heap space" error. The current setup allows the container to use not more than 2GB of RAM. By default, JDK assigns 25 percent of the available RAM (as far as I know).

Do you have any suggestions on how to resolve this properly? We could increase the container memory limit but we have rather limited resources.

Thank you in advance.

**Comment by phax** [comment]
Hi Jonas, I can't give you a real good number, as this depends on a lot of parameters. However, I suggest to use at least 4GB. Especially the boilerplate part for Java applications in Docker is quite high.
Additionally I could modify the code for chaning the endpoint URLs to use less caching and instead iterate all elements twice. That means it takes longer but uses less memory. The same can be applied to "bulk certificate update".

What do you think?

**Comment by phax** [comment]
I made some changes and it would be great if you could test, if 7.1.5-SNAPSHOT version improves things, it would be nice.

**Comment by JonasZal** [comment]
Thanks, Philip for the response.
We have added 2GB to the container. It's a production SMP serving the whole country's public sector, so it was easy to convince admins to add resources. I'll check the 7.1.5 version on the ACC environment. I have there ~4000 service groups as well.
When it comes to performance vs container resources, my vote goes to performance. Moreover, you can always play with CATALINA_OPTS by changing -Xms and -Xmx. I'm not sure if I can pass them without breaking the initial environment value set in the Dockerfile.

**Comment by phax** [comment]
Okay thanks Jonas. I will release the 7.1.5 version asap, as these improvements may be relevant to others as well. And I may add the recommendation to run the SMP with at least 4GB of RAM.... :)

**Comment by phax** [comment]
I made some modifications to the 7.1.5 default Docker images to include this line:
```
ENV CATALINA_OPTS="$CATALINA_OPTS -XX:InitialRAMPercentage=10 -XX:MinRAMPercentage=50 -XX:MaxRAMPercentage=80"
```
This should increase the memory amount accessible to Tomcat from 50% to 80% and therefore increase the efficiency.
Additionally I added the suggestion to use at least 4GB of RAM.

Now building 7.1.5 and therefore closing this issue for now. Thanks for bringing it up.
